File: cucumber-expressions/java/src/test/java/io/cucumber/cucumberexpressions/CustomTransformTest.java
package io.cucumber.cucumberexpressions;

import org.junit.Before;
import org.junit.Test;

import java.lang.reflect.Type;
import java.util.Collections;
import java.util.Locale;

import static java.util.regex.Pattern.compile;
import static org.junit.Assert.assertEquals;

public class CustomTransformTest {
    public static class Color {
        public final String name;

        /// [color-constructor]
        public Color(String name) {
            this.name = name;
        }
        /// [color-constructor]

        @Override
        public int hashCode() {
            return name.hashCode();
        }

        @Override
        public boolean equals(Object obj) {
            return ((Color) obj).name.equals(name);
        }
    }

    private TransformLookup transformLookup = new TransformLookup(Locale.ENGLISH);

    @Before
    public void create_transform() {
        /// [add-color-transform]
        transformLookup.addTransform(new SimpleTransform<>(
                "color",
                Color.class,
                "red|blue|yellow",
                new Function<String, Color>() {
                    @Override
                    public Color apply(String name) {
                        return new Color(name);
                    }
                }
        ));
        /// [add-color-transform]
    }

    @Test
    public void transforms_CucumberExpression_arguments_with_expression_type() {
        Expression expression = new CucumberExpression("I have a {color:color} ball", Collections.<Type>emptyList(), transformLookup);
        Object transformedArgumentValue = expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals(new Color("red"), transformedArgumentValue);
    }

    @Test
    public void transforms_CucumberExpression_arguments_with_explicit_type() {
        Expression expression = new CucumberExpression("I have a {color} ball", Collections.<Type>singletonList(Color.class), transformLookup);
        Object transformedArgumentValue = expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals(new Color("red"), transformedArgumentValue);
    }

    @Test
    public void transforms_CucumberExpression_arguments_using_argument_name_as_type() {
        Expression expression = new CucumberExpression("I have a {color} ball", Collections.<Type>emptyList(), transformLookup);
        Object transformedArgumentValue = expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals(new Color("red"), transformedArgumentValue);
    }

    @Test
    public void transforms_CucumberExpression_arguments_with_explicit_type_using_constructor_directly() {
        Expression expression = new CucumberExpression("I have a {color} ball", Collections.<Type>singletonList(Color.class), new TransformLookup(Locale.ENGLISH));
        Color transformedArgumentValue = (Color) expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals("red", transformedArgumentValue.name);
    }

    ///// RegularExpression

    @Test
    public void transforms_RegularExpression_arguments_with_explicit_type() {
        Expression expression = new RegularExpression(compile("I have a (red|blue|yellow) ball"), Collections.<Type>singletonList(Color.class), transformLookup);
        Object transformedArgumentValue = expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals(new Color("red"), transformedArgumentValue);
    }

    @Test
    public void transforms_RegularExpression_arguments_without_explicit_type() {
        Expression expression = new RegularExpression(compile("I have a (red|blue|yellow) ball"), Collections.<Type>emptyList(), transformLookup);
        Object transformedArgumentValue = expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals(new Color("red"), transformedArgumentValue);
    }

    @Test
    public void transforms_RegularExpression_arguments_with_explicit_type_using_constructor_directly() {
        Expression expression = new RegularExpression(compile("I have a (red|blue|yellow) ball"), Collections.<Type>singletonList(Color.class), new TransformLookup(Locale.ENGLISH));
        Color transformedArgumentValue = (Color) expression.match("I have a red ball").get(0).getTransformedValue();
        assertEquals("red", transformedArgumentValue.name);
    }


}

File: cucumber-expressions/javascript/test/custom_transform_test.js
/* eslint-env mocha */
'use strict'

const assert = require('assert')
const CucumberExpression = require('../src/cucumber_expression')
const RegularExpression = require('../src/regular_expression')
const TransformLookup = require('../src/transform_lookup')
const Transform = require('../src/transform')

class Color {
  /// [color-constructor]
  constructor(name) {
    this.name = name
  }
  /// [color-constructor]
}

describe('Custom transform', () => {
  let transformLookup

  beforeEach(() => {
    /// [add-color-transform]
    transformLookup = new TransformLookup()
    transformLookup.addTransform(new Transform(
      'color',
      Color,
      ['red|blue|yellow', '(?:dark|light) (?:red|blue|yellow)'],
      s => new Color(s)
    ))
    /// [add-color-transform]
  })

  describe(CucumberExpression.name, () => {
    it("transforms arguments with expression type", () => {
      const expression = new CucumberExpression("I have a {color:color} ball", [], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    it("transforms arguments with expression type", () => {
      const expression = new CucumberExpression("I have a {color:color} ball", [], transformLookup)
      const transformedArgumentValue = expression.match("I have a dark red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "dark red")
    })

    it("transforms arguments with explicit type", () => {
      const expression = new CucumberExpression("I have a {color} ball", [Color], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    it("transforms arguments using argument name as type", () => {
      const expression = new CucumberExpression("I have a {color} ball", [], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    it("transforms arguments with explicit type using constructor directly", () => {
      const expression = new CucumberExpression("I have a {color} ball", [Color], new TransformLookup())
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    // JavaScript-specific

    it("transforms arguments with explicit type name", () => {
      const expression = new CucumberExpression("I have a {color} ball", ['color'], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })
  })

  describe(RegularExpression.name, () => {
    it("transforms arguments with explicit type", () => {
      const expression = new RegularExpression(/I have a (red|blue|yellow) ball/, [Color], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    it("transforms arguments without explicit type", () => {
      const expression = new RegularExpression(/I have a (red|blue|yellow) ball/, [], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    it("transforms arguments with explicit type using constructor directly", () => {
      const expression = new RegularExpression(/I have a (red|blue|yellow) ball/, [Color], new TransformLookup())
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })

    // JavaScript-specific (specifying type as string)
    it("transforms arguments with explicit type name", () => {
      const expression = new RegularExpression(/I have a (red|blue|yellow) ball/, ['color'], transformLookup)
      const transformedArgumentValue = expression.match("I have a red ball")[0].transformedValue
      assert.equal(transformedArgumentValue.name, "red")
    })
  })
})
